File: gateways/go/internal/gateway/oidc.go
package gateway

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/coreos/go-oidc/v3/oidc"
	"golang.org/x/oauth2"

	"github.com/kamkie/stackverse/gateways/go/internal/config"
	"github.com/kamkie/stackverse/gateways/go/internal/logx"
	"github.com/kamkie/stackverse/gateways/go/internal/session"
)

var (
	errRefreshRejected = errors.New("refresh rejected by idp")
	errIDPUnavailable  = errors.New("idp unavailable")
)

type oidcClient struct {
	config     config.Config
	oauth      oauth2.Config
	logoutURL  string
	verifier   *oidc.IDTokenVerifier
	httpClient *http.Client
	logger     *slog.Logger
}

type idTokenClaims struct {
	Subject           string `json:"sub"`
	PreferredUsername string `json:"preferred_username"`
}

func newOIDCClient(ctx context.Context, cfg config.Config, httpClient *http.Client, logger *slog.Logger) *oidcClient {
	authURL := cfg.OIDCIssuerURI + "/protocol/openid-connect/auth"
	internal := cfg.OIDCInternalIssuerURI
	tokenURL := internal + "/protocol/openid-connect/token"
	jwksURL := internal + "/protocol/openid-connect/certs"
	logoutURL := internal + "/protocol/openid-connect/logout"

	clientCtx := oidc.ClientContext(ctx, httpClient)
	keySet := oidc.NewRemoteKeySet(clientCtx, jwksURL)
	verifier := oidc.NewVerifier(cfg.OIDCIssuerURI, keySet, &oidc.Config{ClientID: cfg.OIDCClientID})

	return &oidcClient{
		config: cfg,
		oauth: oauth2.Config{
			ClientID:     cfg.OIDCClientID,
			ClientSecret: cfg.OIDCClientSecret,
			RedirectURL:  cfg.RedirectURI(),
			Scopes:       []string{oidc.ScopeOpenID, "profile", "email"},
			Endpoint: oauth2.Endpoint{
				AuthURL:  authURL,
				TokenURL: tokenURL,
			},
		},
		logoutURL:  logoutURL,
		verifier:   verifier,
		httpClient: httpClient,
		logger:     logger,
	}
}

func (c *oidcClient) authCodeURL(state, verifier string) string {
	return c.oauth.AuthCodeURL(state,
		oauth2.AccessTypeOffline,
		oauth2.SetAuthURLParam("code_challenge", codeChallenge(verifier)),
		oauth2.SetAuthURLParam("code_challenge_method", "S256"),
	)
}

func (c *oidcClient) exchange(ctx context.Context, code, verifier string) (session.Data, error) {
	ctx = context.WithValue(ctx, oauth2.HTTPClient, c.httpClient)
	token, err := c.oauth.Exchange(ctx, code, oauth2.SetAuthURLParam("code_verifier", verifier))
	if err != nil {
		return session.Data{}, err
	}
	rawIDToken, _ := token.Extra("id_token").(string)
	if rawIDToken == "" {
		return session.Data{}, errors.New("id_token missing")
	}
	idToken, err := c.verifier.Verify(ctx, rawIDToken)
	if err != nil {
		return session.Data{}, err
	}
	var claims idTokenClaims
	if err := idToken.Claims(&claims); err != nil {
		return session.Data{}, err
	}
	username := claims.PreferredUsername
	if username == "" {
		username = claims.Subject
	}
	if username == "" || token.AccessToken == "" || token.RefreshToken == "" {
		return session.Data{}, errors.New("token response missing required fields")
	}
	now := time.Now().UTC()
	return session.Data{
		Username:     username,
		AccessToken:  token.AccessToken,
		RefreshToken: token.RefreshToken,
		IDToken:      rawIDToken,
		ExpiresAt:    token.Expiry.UTC(),
		CreatedAt:    now,
		UpdatedAt:    now,
	}, nil
}

func (c *oidcClient) refresh(ctx context.Context, data session.Data) (session.Data, error) {
	form := url.Values{
		"grant_type":    {"refresh_token"},
		"refresh_token": {data.RefreshToken},
		"client_id":     {c.config.OIDCClientID},
		"client_secret": {c.config.OIDCClientSecret},
	}
	start := time.Now()
	request, err := http.NewRequestWithContext(ctx, http.MethodPost, c.oauth.Endpoint.TokenURL, strings.NewReader(form.Encode()))
	if err != nil {
		return session.Data{}, err
	}
	request.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	response, err := c.httpClient.Do(request)
	if err != nil {
		logx.Event(ctx, c.logger, slog.LevelError, "dependency_call_failed", "failure",
			"Keycloak was unreachable during token refresh; the session is kept",
			slog.String("dependency", "keycloak"),
			slog.Int64("duration_ms", time.Since(start).Milliseconds()),
			slog.String("error_code", "idp_unreachable"))
		return session.Data{}, errIDPUnavailable
	}
	defer response.Body.Close()

	if response.StatusCode < 200 || response.StatusCode >= 300 {
		if response.StatusCode == http.StatusBadRequest || response.StatusCode == http.StatusUnauthorized {
			logx.Event(ctx, c.logger, slog.LevelWarn, "token_refresh_failed", "failure",
				"Token refresh rejected by the IdP; treating the session as expired",
				slog.String("error_code", "idp_rejected"),
				slog.Int("idp_status", response.StatusCode))
			return session.Data{}, errRefreshRejected
		}
		logx.Event(ctx, c.logger, slog.LevelError, "dependency_call_failed", "failure",
			"Keycloak failed during token refresh; the session is kept",
			slog.String("dependency", "keycloak"),
			slog.Int64("duration_ms", time.Since(start).Milliseconds()),
			slog.String("error_code", "idp_status_"+strconv.Itoa(response.StatusCode)))
		return session.Data{}, errIDPUnavailable
	}

	var payload struct {
		AccessToken  string `json:"access_token"`
		RefreshToken string `json:"refresh_token"`
		ExpiresIn    int    `json:"expires_in"`
		IDToken      string `json:"id_token"`
	}
	if err := json.NewDecoder(response.Body).Decode(&payload); err != nil {
		logx.Event(ctx, c.logger, slog.LevelError, "dependency_call_failed", "failure",
			"Keycloak returned an unreadable token response; the session is kept",
			slog.String("dependency", "keycloak"),
			slog.Int64("duration_ms", time.Since(start).Milliseconds()),
			slog.String("error_code", "idp_bad_response"))
		return session.Data{}, errIDPUnavailable
	}
	if payload.AccessToken == "" {
		logx.Event(ctx, c.logger, slog.LevelError, "dependency_call_failed", "failure",
			"Keycloak returned a token response without an access token; the session is kept",
			slog.String("dependency", "keycloak"),
			slog.Int64("duration_ms", time.Since(start).Milliseconds()),
			slog.String("error_code", "idp_bad_response"))
		return session.Data{}, errIDPUnavailable
	}
	if payload.RefreshToken == "" {
		payload.RefreshToken = data.RefreshToken
	}
	if payload.ExpiresIn <= 0 {
		payload.ExpiresIn = 300
	}
	if payload.IDToken == "" {
		payload.IDToken = data.IDToken
	}
	data.AccessToken = payload.AccessToken
	data.RefreshToken = payload.RefreshToken
	data.IDToken = payload.IDToken
	data.ExpiresAt = time.Now().UTC().Add(time.Duration(payload.ExpiresIn) * time.Second)
	data.UpdatedAt = time.Now().UTC()
	return data, nil
}

func (c *oidcClient) logout(ctx context.Context, refreshToken string) {
	if refreshToken == "" {
		return
	}
	form := url.Values{
		"client_id":     {c.config.OIDCClientID},
		"client_secret": {c.config.OIDCClientSecret},
		"refresh_token": {refreshToken},
	}
	request, err := http.NewRequestWithContext(ctx, http.MethodPost, c.logoutURL, strings.NewReader(form.Encode()))
	if err != nil {
		return
	}
	request.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	response, err := c.httpClient.Do(request)
	if err != nil {
		logx.Event(ctx, c.logger, slog.LevelWarn, "idp_logout_failed", "failure",
			"IdP logout failed; local session destroyed anyway",
			slog.String("error_code", "idp_unreachable"))
		return
	}
	defer response.Body.Close()
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		logx.Event(ctx, c.logger, slog.LevelWarn, "idp_logout_failed", "failure",
			fmt.Sprintf("IdP logout returned %d; local session destroyed anyway", response.StatusCode),
			slog.String("error_code", "idp_rejected"),
			slog.Int("idp_status", response.StatusCode))
	}
}

File: gateways/rust/src/oidc.rs
use std::collections::HashMap;
use std::sync::Arc;
use std::time::{Duration, Instant};

use anyhow::{Context, anyhow};
use base64::Engine;
use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use chrono::{TimeDelta, Utc};
use jsonwebtoken::{Algorithm, DecodingKey, Validation, decode, decode_header};
use serde::Deserialize;
use sha2::{Digest, Sha256};
use thiserror::Error;
use tokio::sync::Mutex;
use url::Url;

use crate::config::Config;
use crate::session::SessionData;

const JWKS_REFRESH_COOLDOWN: Duration = Duration::from_secs(30);

#[derive(Clone)]
pub struct OidcClient {
    config: Arc<Config>,
    http: reqwest::Client,
    authorization_url: String,
    token_url: String,
    logout_url: String,
    jwks_url: String,
    jwks: Arc<Mutex<JwksState>>,
}

#[derive(Default)]
struct JwksState {
    keys: HashMap<String, DecodingKey>,
    last_refresh: Option<Instant>,
}

#[derive(Deserialize)]
struct JwksDocument {
    keys: Vec<Jwk>,
}

#[derive(Deserialize)]
struct Jwk {
    kid: String,
    kty: String,
    n: String,
    e: String,
}

#[derive(Deserialize)]
struct TokenResponse {
    access_token: String,
    #[serde(default)]
    refresh_token: Option<String>,
    #[serde(default)]
    id_token: Option<String>,
    #[serde(default)]
    expires_in: Option<i64>,
}

#[derive(Debug, Deserialize)]
struct IdClaims {
    sub: String,
    #[serde(default)]
    preferred_username: Option<String>,
}

#[derive(Debug, Error)]
pub enum RefreshError {
    #[error("refresh rejected by idp")]
    Rejected,
    #[error("idp unavailable")]
    Unavailable,
}

impl OidcClient {
    pub fn new(config: Arc<Config>, http: reqwest::Client) -> Self {
        let issuer = config.oidc_issuer_uri.clone();
        let internal = config.oidc_internal_issuer_uri.clone();
        Self {
            config,
            http,
            authorization_url: format!("{issuer}/protocol/openid-connect/auth"),
            token_url: format!("{internal}/protocol/openid-connect/token"),
            logout_url: format!("{internal}/protocol/openid-connect/logout"),
            jwks_url: format!("{internal}/protocol/openid-connect/certs"),
            jwks: Arc::new(Mutex::new(JwksState::default())),
        }
    }

    pub fn authorization_redirect(&self, state: &str, verifier: &str) -> anyhow::Result<String> {
        let mut url = Url::parse(&self.authorization_url)?;
        url.query_pairs_mut()
            .append_pair("response_type", "code")
            .append_pair("client_id", &self.config.oidc_client_id)
            .append_pair("redirect_uri", &self.config.redirect_uri())
            .append_pair("scope", "openid profile email")
            .append_pair("state", state)
            .append_pair("code_challenge", &code_challenge(verifier))
            .append_pair("code_challenge_method", "S256");
        Ok(url.to_string())
    }

    pub async fn exchange(&self, code: &str, verifier: &str) -> anyhow::Result<SessionData> {
        let response = self
            .http
            .post(&self.token_url)
            .form(&[
                ("grant_type", "authorization_code"),
                ("code", code),
                ("redirect_uri", &self.config.redirect_uri()),
                ("client_id", &self.config.oidc_client_id),
                ("client_secret", &self.config.oidc_client_secret),
                ("code_verifier", verifier),
            ])
            .send()
            .await
            .context("exchange authorization code")?;
        if !response.status().is_success() {
            return Err(anyhow!("token endpoint returned {}", response.status()));
        }
        let payload = response.json::<TokenResponse>().await?;
        let raw_id_token = payload
            .id_token
            .context("token response missing id_token")?;
        let username = self.validate_id_token(&raw_id_token).await?;
        if payload.access_token.is_empty() {
            return Err(anyhow!("token response missing access token"));
        }
        let refresh_token = payload
            .refresh_token
            .filter(|value| !value.is_empty())
            .context("token response missing refresh token")?;
        let now = Utc::now();
        Ok(SessionData {
            username,
            access_token: payload.access_token,
            refresh_token,
            id_token: raw_id_token,
            expires_at: now + TimeDelta::seconds(payload.expires_in.unwrap_or(300).max(1)),
            created_at: now,
            updated_at: now,
        })
    }

    pub async fn refresh(&self, mut data: SessionData) -> Result<SessionData, RefreshError> {
        let started = Instant::now();
        let response = self
            .http
            .post(&self.token_url)
            .form(&[
                ("grant_type", "refresh_token"),
                ("refresh_token", data.refresh_token.as_str()),
                ("client_id", &self.config.oidc_client_id),
                ("client_secret", &self.config.oidc_client_secret),
            ])
            .send()
            .await
            .map_err(|err| {
                tracing::error!(
                    event = "dependency_call_failed",
                    outcome = "failure",
                    dependency = "keycloak",
                    duration_ms = started.elapsed().as_millis() as i64,
                    error_code = "idp_unreachable",
                    error = %err,
                    "Keycloak was unreachable during token refresh; the session is kept"
                );
                RefreshError::Unavailable
            })?;

        if !response.status().is_success() {
            let status = response.status();
            if status == reqwest::StatusCode::BAD_REQUEST
                || status == reqwest::StatusCode::UNAUTHORIZED
            {
                tracing::warn!(
                    event = "token_refresh_failed",
                    outcome = "failure",
                    error_code = "idp_rejected",
                    idp_status = status.as_u16(),
                    "Token refresh rejected by the IdP; treating the session as expired"
                );
                return Err(RefreshError::Rejected);
            }
            tracing::error!(
                event = "dependency_call_failed",
                outcome = "failure",
                dependency = "keycloak",
                duration_ms = started.elapsed().as_millis() as i64,
                error_code = %format!("idp_status_{}", status.as_u16()),
                "Keycloak failed during token refresh; the session is kept"
            );
            return Err(RefreshError::Unavailable);
        }

        let payload = response.json::<TokenResponse>().await.map_err(|err| {
            tracing::error!(
                event = "dependency_call_failed",
                outcome = "failure",
                dependency = "keycloak",
                duration_ms = started.elapsed().as_millis() as i64,
                error_code = "idp_bad_response",
                error = %err,
                "Keycloak returned an unreadable token response; the session is kept"
            );
            RefreshError::Unavailable
        })?;

        if payload.access_token.is_empty() {
            tracing::error!(
                event = "dependency_call_failed",
                outcome = "failure",
                dependency = "keycloak",
                duration_ms = started.elapsed().as_millis() as i64,
                error_code = "idp_bad_response",
                "Keycloak returned a token response without an access token; the session is kept"
            );
            return Err(RefreshError::Unavailable);
        }

        let now = Utc::now();
        data.access_token = payload.access_token;
        if let Some(refresh_token) = payload.refresh_token.filter(|value| !value.is_empty()) {
            data.refresh_token = refresh_token;
        }
        if let Some(id_token) = payload.id_token.filter(|value| !value.is_empty()) {
            data.id_token = id_token;
        }
        data.expires_at = now + TimeDelta::seconds(payload.expires_in.unwrap_or(300).max(1));
        data.updated_at = now;
        Ok(data)
    }

    pub async fn logout(&self, refresh_token: &str) {
        if refresh_token.is_empty() {
            return;
        }
        let response = self
            .http
            .post(&self.logout_url)
            .form(&[
                ("client_id", self.config.oidc_client_id.as_str()),
                ("client_secret", self.config.oidc_client_secret.as_str()),
                ("refresh_token", refresh_token),
            ])
            .send()
            .await;
        match response {
            Ok(response) if response.status().is_success() => {}
            Ok(response) => {
                tracing::warn!(
                    event = "idp_logout_failed",
                    outcome = "failure",
                    error_code = "idp_rejected",
                    idp_status = response.status().as_u16(),
                    "IdP logout returned a non-success status; local session destroyed anyway"
                );
            }
            Err(err) => {
                tracing::warn!(
                    event = "idp_logout_failed",
                    outcome = "failure",
                    error_code = "idp_unreachable",
                    error = %err,
                    "IdP logout failed; local session destroyed anyway"
                );
            }
        }
    }

    async fn validate_id_token(&self, raw: &str) -> anyhow::Result<String> {
        let header = decode_header(raw)?;
        let kid = header.kid.context("ID token has no kid")?;
        let key = self.key(&kid).await?;
        let mut validation = Validation::new(Algorithm::RS256);
        validation.set_issuer(std::slice::from_ref(&self.config.oidc_issuer_uri));
        validation.set_audience(std::slice::from_ref(&self.config.oidc_client_id));
        let token = decode::<IdClaims>(raw, &key, &validation)?;
        let username = token
            .claims
            .preferred_username
            .filter(|value| !value.is_empty())
            .unwrap_or(token.claims.sub);
        if username.is_empty() {
            return Err(anyhow!("ID token missing subject"));
        }
        Ok(username)
    }

    async fn key(&self, kid: &str) -> anyhow::Result<DecodingKey> {
        let mut state = self.jwks.lock().await;
        if let Some(key) = state.keys.get(kid) {
            return Ok(key.clone());
        }
        if state
            .last_refresh
            .is_some_and(|last| last.elapsed() < JWKS_REFRESH_COOLDOWN)
        {
            return Err(anyhow!("unknown signing key {kid}"));
        }
        self.refresh_keys_locked(&mut state).await?;
        state
            .keys
            .get(kid)
            .cloned()
            .ok_or_else(|| anyhow!("unknown signing key {kid}"))
    }

    async fn refresh_keys_locked(&self, state: &mut JwksState) -> anyhow::Result<()> {
        state.last_refresh = Some(Instant::now());
        let started = Instant::now();
        let document = self
            .http
            .get(&self.jwks_url)
            .send()
            .await
            .and_then(|response| response.error_for_status())
            .context("fetch JWKS")
            .map_err(|err| {
                tracing::error!(
                    event = "dependency_call_failed",
                    outcome = "failure",
                    dependency = "keycloak",
                    duration_ms = started.elapsed().as_millis() as i64,
                    error_code = "jwks_fetch_failed",
                    error = %err,
                    "Fetching the JWKS from the IdP failed"
                );
                err
            })?
            .json::<JwksDocument>()
            .await?;
        let mut keys = HashMap::new();
        for jwk in document.keys {
            if jwk.kty != "RSA" {
                continue;
            }
            if let Ok(key) = DecodingKey::from_rsa_components(&jwk.n, &jwk.e) {
                keys.insert(jwk.kid, key);
            }
        }
        if keys.is_empty() {
            return Err(anyhow!(
                "the JWKS document contains no usable RSA signing keys"
            ));
        }
        state.keys = keys;
        Ok(())
    }
}

pub fn new_opaque_id() -> String {
    let bytes: [u8; 32] = rand::random();
    bytes.iter().map(|byte| format!("{byte:02x}")).collect()
}

pub fn new_code_verifier() -> String {
    let bytes: [u8; 32] = rand::random();
    URL_SAFE_NO_PAD.encode(bytes)
}

fn code_challenge(verifier: &str) -> String {
    URL_SAFE_NO_PAD.encode(Sha256::digest(verifier.as_bytes()))
}
